File: src/searcher.py
# This is a class defining a searching object to query various DNS servers

import requests
import json
import socket


class Searcher:
    def __init__(self, target):
        self._target = target
        self._results = []

    @property
    def results(self):
        return self._results

    def googledns(self):
        # Query Google DNS 4 times and append the answers to the results list
        for i in range(4):
            # This is only for debug purposes:
            #print('Google DNS attempt: ' + str(i))
            r = requests.get('https://dns.google.com/resolve?name=' + self._target + '&type=A')

            # Check HTTP status, and append answers to the results attribute of the searcher object
            if r.ok:
                jdat = json.loads(r.text)
                for o in jdat['Answer']:
                    if o['type'] == 1:
                        self._results.append(o['data'])

    def cloudflaredns(self):
        # Query Cloudflare DNS 4 times and append the answers to the results list
        for i in range(4):
            # This is only for debug purposes:
            #print('Cloudflare DNS attempt: ' + str(i))
            headers = {'accept': 'application/dns-json'}
            r = requests.get('https://cloudflare-dns.com/dns-query?name=' + self._target + '&type=A', headers=headers)

            # Check HTTP status, and append answers to the results attribute of the searcher object
            if r.ok:
                jdat = json.loads(r.text)
                for o in jdat['Answer']:
                    if o['type'] == 1:
                        self._results.append(o['data'])

    def localresolver(self):
        # Query the locally assigned resolver 4 times.
        for i in range(4):
            # For debugging purposes:
            #print('Local DNS attempt: ' + str(i))
            resp = socket.gethostbyname(self._target)
            self._results.append(resp)

File: cmd/ip-hunter/searcher.go
package main

import (
	"encoding/json"
	"fmt"
	"net"
	"net/http"
	"os"
	"time"
)

type searcher struct {
	Target  string
	Count   int
	Results map[string]bool
}

type Payload struct {
	Answer []*PayloadData `json:"answer"`
}

type PayloadData struct {
	Data string `json:"data"`
	Type int    `json:"type"`
}

func (s *searcher) printResults() {
	fmt.Println("\nResults:")
	for k, _ := range s.Results {
		fmt.Println(k)
	}
}

func (s *searcher) uniqueResults(x map[string]bool) {
	for k, _ := range x {
		if !s.Results[k] {
			s.Results[k] = true
		}
	}
}

func (s *searcher) googledns(c chan map[string]bool) {
	req, _ := http.NewRequest("GET", "https://dns.google.com/resolve?name="+s.Target+"&type=A", nil)
	results := getDohIps(req, s.Count)
	c <- results
}

func (s *searcher) cloudflaredns(c chan map[string]bool) {
	req, _ := http.NewRequest("GET", "https://cloudflare-dns.com/dns-query?name="+s.Target+"&type=A", nil)
	req.Header.Set("accept", "application/dns-json")
	results := getDohIps(req, s.Count)
	c <- results
}

func (s *searcher) localresolver(c chan map[string]bool) {
	results := map[string]bool{}
	for i := 1; i <= s.Count; i++ {
		ips, err := net.LookupIP(s.Target)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Could not get IPs: %v\n", err)
			break
		}
		for _, v := range ips {
			// for now only get ipv4 addresses
			if len(v) == 4 {
				if !results[v.String()] {
					results[v.String()] = true
				}
			}
		}
	}
	c <- results
}

func getClient() http.Client {
	timeout := time.Duration(5 * time.Second)
	client := http.Client{
		Timeout: timeout,
	}
	return client
}

func getDohIps(req *http.Request, count int) map[string]bool {
	client := getClient()
	res := map[string]bool{}

	for i := 1; i <= count; i++ {
		resp, err := client.Do(req)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Failed HTTP request to %v\n", req.URL)
			break
		}
		var p Payload
		err = json.NewDecoder(resp.Body).Decode(&p)
		if err != nil {
			fmt.Fprintf(os.Stderr, "[!]Failed to decode DoH JSON results from %v\n", req.URL)
			break
		}
		// Loop through results and add to the map if the value doesn't already exist
		// Also check to ensure we are getting a "type 1" A record response
		for _, v := range p.Answer {
			if v.Type == 1 {
				if !res[v.Data] {
					res[v.Data] = true
				}
			}
		}
	}

	return res
}
